File: MainApp/src/main/java/com/sleticalboy/learning/base/BaseRVAdapter.java
package com.sleticalboy.learning.base;

import android.view.ViewGroup;

import androidx.annotation.NonNull;
import androidx.recyclerview.widget.RecyclerView;

import java.util.ArrayList;
import java.util.Arrays;
import java.util.List;

/**
 * Created on 20-8-23.
 *
 * @author binlee sleticalboy@gmail.com
 */
public abstract class BaseRVAdapter<DATA> extends RecyclerView.Adapter<BaseRVHolder<DATA>> {

    protected final List<DATA> mData;

    public BaseRVAdapter() {
        this(new ArrayList<>());
    }

    public BaseRVAdapter(final List<DATA> data) {
        mData = data == null ? new ArrayList<>() : data;
    }

    public BaseRVAdapter(final DATA[] data) {
        this(data == null ? new ArrayList<>() : new ArrayList<>(Arrays.asList(data)));
    }

    @NonNull
    @Override
    public abstract BaseRVHolder<DATA> onCreateViewHolder(@NonNull final ViewGroup parent, final int viewType);

    @Override
    public final void onBindViewHolder(@NonNull final BaseRVHolder holder, final int position) {
        holder.bindData(getData(position), position);
    }

    private DATA getData(final int position) {
        if (position >= mData.size()) {
            throw new IllegalArgumentException("position >= mData.size(): " + position);
        }
        return mData.get(position);
    }

    public final int addData(DATA data) {
        final int index = mData.indexOf(data);
        if (index < 0) {
            mData.add(data);
        } else {
            mData.set(index, data);
        }
        final int position = index < 0 ? mData.size() - 1 : index;
        notifyItemChanged(position);
        // Log.d(logTag(), "onDeviceScanned() index: $index, device: $device");
        return position;
    }

    @Override
    public int getItemCount() {
        return mData.size();
    }
}

File: MainApp/src/main/java/com/sleticalboy/learning/base/BaseRVAdapter.kt
package com.sleticalboy.learning.base

import android.view.ViewGroup
import androidx.recyclerview.widget.RecyclerView
import java.util.ArrayList

/**
 * Created on 20-8-23.
 *
 * @author binlee sleticalboy@gmail.com
 */
abstract class BaseRVAdapter<DATA> @JvmOverloads constructor(data: List<DATA>? = ArrayList()) :
  RecyclerView.Adapter<BaseRVHolder<DATA>>() {

  protected val mData: MutableList<DATA> = data as MutableList<DATA>? ?: ArrayList()

  constructor(data: Array<DATA>?) : this(data?.toList() ?: ArrayList<DATA>())

  abstract override fun onCreateViewHolder(parent: ViewGroup, viewType: Int): BaseRVHolder<DATA>

  override fun onBindViewHolder(holder: BaseRVHolder<DATA>, position: Int) {
    holder.bindData(getData(position), position)
  }

  private fun getData(position: Int): DATA {
    require(position < mData.size) { "position >= mData.size(): $position" }
    return mData[position]
  }

  fun addData(data: DATA): Int {
    val index = mData.indexOf(data)
    if (index < 0) {
      mData.add(data)
    } else {
      mData[index] = data
    }
    val position = if (index < 0) mData.size - 1 else index
    notifyItemChanged(position)
    // Log.d(logTag(), "onDeviceScanned() index: $index, device: $device");
    return position
  }

  override fun getItemCount(): Int {
    return mData.size
  }

}